docs/config-validation.md

```diff
@@ -0,0 +1,191 @@
+# Configuration Validation
+
+We use a fork of
+[github.com/go-playground/validator](https://github.com/go-playground/validator)
+to enable validation of our configuration files using struct tags. The validator
+is configured to validate the configuration file at component startup if the
+`-validate` command line flag is set. If the supplied configuration file is
+valid, the component will exit with a zero exit code, otherwise it will exit
+with a non-zero exit code and print the validation errors to stderr.
+
+## Struct Tag Tips
+
+You can find the full list of struct tags supported by the validator [here]
+(https://pkg.go.dev/github.com/go-playground/validator/v10#section-documentation).
+The following are some tips for struct tags that are commonly used in our
+configuration files.
+
+### `required`
+
+This only validates that the value is not the data types default zero value. For
+numbers ensures value is not zero. For strings ensures value is not "". For
+slices, maps, pointers, interfaces, channels and functions ensures the value is
+not nil. Note, this does not validate slices, maps, etc., as they are not nil.
+For these fields you should use `min=1` or `gte=1` to validate the supplied
+values are not empty.
+
+### `omitempty`
+
+The validator will always run any present validations unless the field is also
+tagged with `omitempty`. Said another way, if a validation is present, and that
+validation cannot be satisfied by the zero-value of that field type, then that
+field is also (technically) required. However, when `omitempty` is present,
+validations will only run when the condition of a `required_` (`with`,
+`with_all`, `unless`, etc.) tag is met.
+
+### `-`
+This tag is used to ignore a struct or field. It can be useful if you have a
+struct that is nested into various other structs but isn't always required. This
+is the same as saying a struct and all of its fields are optional. We use this
+tag for many config duration and password file struct valued fields which are
+optional in some configs but required in others.
+
+### `structonly`
+
+Very similar to `omitempty`, but used to control whether or not the validator
+will run validations present for the fields of a nested struct. When a nested
+struct is tagged with `structonly` the validations for its fields will only run
+when the condition of a `required_` (`with`, `with_all`, `unless`, etc.) tag is
+met.
```

Comment:
I'm similarly confused here -- either I don't understand what this paragraph is saying, or I don't understand what the upstream validator tag docs are saying (or both!).